nnstreamer/tensor_trainer/tensor_trainer_nntrainer.cc

```diff
@@ -0,0 +1,534 @@
+  ml_logd("num_epochs: %" PRId64 "", num_epochs);
+  ml_logd("Total number of data to be received: %" PRId64 "",
+          total_num_samples);
+  ml_logd("model_config: %s", model_config.c_str());
+  ml_logd("model_save_path: %s", model_save_path.c_str());
+  ml_logd("<leave>");
+}
+
+static int nntrainer_model_destructor(const GstTensorTrainerFramework *fw,
+                                      const GstTensorTrainerProperties *prop,
+                                      void **private_data) {
+  NNTrainer::NNTrainerTrain *nntrainer =
+    static_cast<NNTrainer::NNTrainerTrain *>(*private_data);
+  UNUSED(fw);
+  ml_logd("<called>");
+
+  if (!nntrainer)
+    return -1;
+
+  delete nntrainer;
+  *private_data = NULL;
+  ml_logd("<leave>");
+
+  return 0;
+}
+
+static int
+nntrainer_model_start_training(const GstTensorTrainerFramework *fw,
+                               const GstTensorTrainerProperties *prop,
+                               void *private_data) {
+  NNTrainer::NNTrainerTrain *nntrainer =
+    reinterpret_cast<NNTrainer::NNTrainerTrain *>(private_data);
+  UNUSED(fw);
+
+  ml_logd("<called>");
+  if (!nntrainer) {
+    ml_loge("Failed get nntrainer");
+  }
+  try {
+    std::thread train_thread(nntrainer_thread_func, nntrainer);
+    train_thread.detach();
+  } catch (const std::exception &e) {
+    ml_loge("Error %s, %s", typeid(e).name(), e.what());
+    return -1;
+  }
+  ml_logd("<leave>");
+  return 0;
+}
+
+void NNTrainer::NNTrainerTrain::trainModel() {
+  pid_t pid = getpid();
+  pid_t tid = syscall(SYS_gettid);
+
+  ml_logd("<called>");
+  ml_logd("pid[%d], tid[%d]", pid, tid);
+
+  try {
+    model->setProperty({"epochs=" + std::to_string(num_epochs)});
+  } catch (const std::exception &e) {
+    ml_loge("Error %s, %s", typeid(e).name(), e.what());
+    return;
+  }
+
+  try {
+    model->train();
+    training_loss = model->getTrainingLoss();
+    validation_loss = model->getValidationLoss();
+  } catch (const std::exception &e) {
+    ml_loge("Error %s, %s", typeid(e).name(), e.what());
+    return;
+  }
+  ml_logd("training_loss: %f, validation_loss: %f", training_loss,
+          validation_loss);
+  try {
+    ml_logd("Save_model: %s", model_save_path.c_str());
+    model->save(model_save_path, ml::train::ModelFormat::MODEL_FORMAT_BIN);
+    is_train_complete = TRUE;
+    ml_logd("send train_complete_cond signal");
+    g_cond_signal(train_complete_cond);
+
+  } catch (const std::exception &e) {
+    ml_loge("Error %s, %s", typeid(e).name(), e.what());
+    return;
+  }
+  ml_logd("<leave>");
+}
+
+void NNTrainer::NNTrainerTrain::createModel() {
+  ml_logd("<called>");
+  try {
+    model = ml::train::createModel(ml::train::ModelType::NEURAL_NET);
+  } catch (const std::exception &e) {
+    ml_loge("Error %s, %s", typeid(e).name(), e.what());
+    return;
+  }
+  try {
+
+    model->load(model_config,
+                ml::train::ModelFormat::MODEL_FORMAT_INI_WITH_BIN);
```

Comment:
"ml::train::ModelFormat::MODEL_FORMAT_INI_WITH_BIN" is not the only option nntrainer has. is it intentional using it as static variable or it's just for this PR and will be extend later?

Reply:
Currently I'm only thinking of creating models in the ini.
So it's hardcoded.
I will modify the interface when function is extended.

Reply:
I just leave the comment about the MODEL_FORMAT_INI_WITH_BIN. It load the network model as in model_config and it try to load the binary data which described in ini file as save_path. I wonder if this process is what you want? if it is, then I recommend to leave the comment here saying it load the pretrained binary model as in save_path keyword in ini file.